### PR TITLE
[BACKLOG-18740] In complex add null atoms if only defined at the root proto object

### PR DESCRIPTION
**Pull request by nantunes**
@webdetails/millenniumfalcon please review.

### CODE REVIEW

File: package-res/cdo/complex.js [hunk 1]
```diff
@@ -176,6 +179,39 @@ def
...
+
+            current = def.protoOf(current);
+        }
+
+        return __atomExists(root);
+
+        /* Inner auxiliary function.
+         * @private
+         */
+        function __atomExists(complex) {
```
Comment (dcleao) [inline package-res/cdo/complex.js]:
```js
var dimName = atom.dimension.name;
var atomExists = (def.getOwn(atoms, dimName) === atom);
```

File: package-res/cdo/complex.js [hunk 2]
```diff
@@ -176,6 +179,39 @@ def
...
     },
 
+    /**
+     * Checks if a given atom is only defined at the root proto object.
+     *
+     * @returns {Boolean} True if the atom is defined at the root proto object and
+     *                    it is not overshadowed by any more specific proto.
+     */
+    isAtomRootDefault: function(atom) {
+        var root = def.rootProtoOf(this.atoms);
```
Comment (dcleao) [inline package-res/cdo/complex.js]: Ideally we would traverse the prototype chain only once. Looking at the code of `def.rootProtoOf`, though, I can see why you do it.

File: package-res/cdo/complex.js [hunk 3]
```diff
@@ -89,9 +89,12 @@ def
                     // Need to intern, even if null.
                     atom = ownerDims[dimName].intern(v);
 
-                // Don't add atoms already in base proto object.
-                // (virtual) nulls are already in the root proto object.
-                if(v != null && (!atomsBase || atom !== atomsBase[dimName])) atomsMap[dimName] = atom;
+                // Don't add atoms already in base proto object,
+                // except for nulls only defined at the root proto object.
+                if(atom.value === null && me.isAtomRootDefault(atom) ||
```
Comment (dcleao) [inline package-res/cdo/complex.js]: Please don't make me have to know the operator precedence of `&&` and `||`.

**Comment by nantunes** [comment]
@dcleao code updated.

**Pull request merged**